### ISSUE TITLE
Async function declaration being parsed as expression when followed by certain punctuators

### ISSUE THREAD

**Issue by UltCombo** [6to5/6to5#781]
The following is parsed as an async function declaration:

```js
async function g() {
  return 43;
}
```

Now if we add a line break followed by an array literal after the declaration, it fails to parse as it treats the array literal as if it were a property access:

```js
async function g() {
  return 43;
}
[];
```

Same behavior for the `()` punctuators, which are being wrongly treated as a function call:

```js
async function g() {
  return 43;
}
(1);
```

**Comment by UltCombo** [comment]
In fact, the line break shouldn't make any difference here.

```js
// function declaration followed by array literal
async function g() {
  return 43;
}[];

// function declaration followed by number literal inside grouping operator
async function g() {
  return 43;
}(1);
```

That is, taking in consideration the regular function declaration parsing rules which I believe should also apply to async function declarations.

**Comment by UltCombo** [comment]
The temporary workaround is to add a semicolon after the declaration.

**Comment by sebmck** [comment]
Expression/statement ambiguity. I'll rewrite the async function parser.

**Comment by sebmck** [comment]
Fixed. Need to figure out how to do `async.foo();` etc without lookaheads.

**Comment by UltCombo** [comment]
Wow, impressive!
It is almost unbelievable how you can fix these things before I can barely get to the corresponding acorn-6to5 parser function. :stuck_out_tongue:

**Comment by sebmck** [comment]
Most of the time it's easy because I already know what the bug is when I hear the problem :stuck_out_tongue_winking_eye:

**Comment by UltCombo** [comment]
Makes sense hahah. Thanks for the quick feedback & fix, as always. :smile:

**Comment by sebmck** [comment]
Superseded by 6to5/6to5#781

**Comment by UltCombo** [comment]
I guess you meant https://github.com/6to5/acorn-6to5/issues/27 ?

**Comment by sebmck** [comment]
@UltCombo Whoops, yeah... copy+paste fail.

**Comment by sebmck** [comment]
Rewrote the async function parsing. Released as of 4.4.4.

**Comment by UltCombo** [comment]
Nice, thanks!
